### ISSUE TITLE
Make TFA infer concrete type for Set()/Map() as with map literals

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#37861]
For this code:
```dart
main() {
  print(globalSet);
  print(globalMap);
  print(globalMapLiteral);
}

Set globalSet = Set();
Map globalMap = Map();
Map globalMapLiteral = {};
```
TFA results in 
```
library from "file:///test.dart" as tes {

  static field core::Set* globalSet = col::LinkedHashSet::•<dynamic>();

  static field core::Map* globalMap = core::Map::•<dynamic, dynamic>();

  [@vm.inferred-type.metadata=dart.collection::_InternalLinkedHashMap?]
  static field core::Map* globalMapLiteral = <dynamic, dynamic>{};

  static method main() → dynamic {
    core::print(tes::globalSet);
    core::print(tes::globalMap);
    core::print([@vm.inferred-type.metadata=dart.collection::_InternalLinkedHashMap?] tes::globalMapLiteral);
  }
}
```

The `LinkedHashSet` constructor looks likes this:
```dart
  factory LinkedHashSet({bool equals(E e1, E e2),int hashCode(E e),bool isValidKey(potentialKey)}) {
    if (isValidKey == null) {
      if (hashCode == null) {
        if (equals == null) {
          return new _CompactLinkedHashSet<E>();
        }
        ...
      } else { ... }
    } else { ... }
    return new _CompactLinkedCustomHashSet<E>(equals, hashCode, isValidKey);
  }
```

/cc @alexmarkov

**Comment by lfkdsk** [comment]
1

**Comment by lfkdsk** [comment]
@mkustermann Hi, I try to impl this issus by add code in `SummaryCollector.visitStaticInvocation` ：
``` dart
  @override
  TypeExpr visitStaticInvocation(StaticInvocation node) {
    final args = _visitArguments(null, node.arguments,
        passTypeArguments: node.target.isFactory);
    final target = node.target;
    final coreTypes = _environment.coreTypes;
    final Procedure linkedHashSet = coreTypes.index.getMember('dart:collection', 'LinkedHashSet', '');
    if (target == linkedHashSet) {
        return _translator.instantiateConcreteType(
            _entryPointsListener
                .addAllocatedClass(coreTypes.index.getClass('dart:collection', '_CompactLinkedHashSet')),
          node.arguments.types,
        );
    }
   ...
```
Found Set's StaticInvocation and infer to its realy extension type just like `MapLiteral`.
Set's create StaticInvocation could be found correctly, but I got error trace like follow: 

``` dart
  Assertion failed. Details: LinkedHashSet.

  package:vm/transformations/type_flow/utils.dart 35:5           assertx
  package:vm/transformations/type_flow/transformer.dart 1258:5   _TreeShakerPass1.visitStaticInvocation
  package:kernel/ast.dart 4091:44                                StaticInvocation.accept
  package:kernel/ast.dart 1877:33                                Field.transformChildren
  package:vm/transformations/type_flow/transformer.dart 1061:16  _TreeShakerPass1.visitField

...
```

It seems like `LinkedHashSet` will be marked as unreachable in `TreeShaking` Pass. 
Is this the right solution? Or do you have any ideas or suggestions for solving this mistake?

**Comment by mkustermann** [comment]
I'll be OOO for some days - maybe @alexmarkov can help you with this.

**Comment by alexmarkov** [comment]
@lfkdsk You're changing how TFA summaries are built, but not the original kernel AST. This breaks correspondence between call sites in the kernel AST and information about call sites maintained by TFA. When tree shaker inspects a `StaticInvocation` AST node, it finds that it is reachable but target of that invocation is not, which is not correct.

I think the better way of doing this would be to specialize `Map()` / `Set()` (or `LinkedHashMap()` / `LinkedHashSet()`) constructor invocations before TFA. We do similar things for list constructors in https://github.com/dart-lang/sdk/blob/master/pkg/vm/lib/transformations/list_factory_specializer.dart

More generally, we could handle this case and other similar cases by adding control statements (something like `IfNull`) to TFA summaries and infer more precise types even without constructor specialization. But this would be much more involved change, which may also negatively affect AOT compilation time.

**Comment by lfkdsk** [comment]
@alexmarkov @mkustermann  
Thanks for the help, I previously thought I could change the TFA results directly by modifying the summary. 
I will first try to implement the original Kernel modifications using the first option (just like `list_factory_specializer.dart`).

**Comment by lfkdsk** [comment]
@alexmarkov I've finished the Map side specializer, in this PR https://github.com/dart-lang/sdk/pull/42943

**Comment by alexmarkov** [comment]
Thank you, @lfkdsk for the contribution. The change landed in 4ef577d50e8bc9240a45d808ef3cc8e1407ed09b.
